[knowledge-base] Add Moderator Single Game Guide Management Backend

yizou275 opened the pull request:
This PR is for issue #794, and supports the feature that a moderator can view a list of approved guides for a specific game and select one to publish.

## Description
The implemented feature includes the following:
1. "ModeratorListByGame" page, displays a list of games with their linked published guide entry (N/A if no guides are published), and provide link to "ModeratorSingleGame" page; show "No games yet" if there is no game
2. "ModeratorSingleGame" page, displays a list of approved guides for a specific game, where moderator can publish, switch, or unpublish; show "No approved guides yet" if there is no approved guides corresponding to the game
3. Allow download for moderators (besides authors of the guide), and include download link in "ModeratorSingleGame" page (future if we have a preview page, can rather link it to the preview page)

## File Changes
Specifically, the files I change are:
- `views.py`: add views for both pages and for managing publishing/unpublishing in database, allow moderator to access `DownloadGuide` view as well
- `urls.py`: add urls according to the views
- `moderator_game_list.html`: template for "ModeratorListByGame" page
- `moderator_single_game.html`: template for "ModeratorSingleGame" page
- `moderator_pending_guides.html`: add a link from moderator view to "ModeratorListByGame" page

## Testing
To test it, first pull this branch, load the fixture according to `knowledge-base/fixtures/README.md`, and then runserver.
1. Log in as Alex. Go to ModeratorView page, and clicks on the "Manage Published Guides". Should arrive at "ModeratorListByGame" page, looking like this:
<img width="1123" alt="Screenshot 2025-05-15 at 10 35 57 PM" src="https://github.com/user-attachments/assets/bfc73f06-0429-4eb6-8663-46cdaac14506" />

2. First check whether "view" link leads to the guide detail page. Then come back and check "manage" link. Should lead to the "ModeratorSingleGame" page, looking like the following for Go Game. Confirm that only guide with ID=2 is included on the page, NOT guide ID=6 (also for Go Game, but rejected). Check the download link works.
<img width="1081" alt="Screenshot 2025-05-15 at 10 40 30 PM" src="https://github.com/user-attachments/assets/f3cb1745-3716-4152-8600-b84ebbf6824b" />

3. Hit the unpublish button. Confirm from admin interface that the `published_guide_id` for Go Game field becomes none. (Can also check the knowledge base landing page or guide detail page)

4. Go to the "ModeratorSingleGame" page for Chess. Confirm it looks like the following. Similarly, check whether "switch to publish" update the database accordingly.

<img width="1062" alt="Screenshot 2025-05-15 at 10 37 32 PM" src="https://github.com/user-attachments/assets/0b4fc601-4d8a-4144-97c7-4b0015c64780" />

5. Check whether the sorting works on the "ModeratorSingleGame" page.
6. Go to the same page for Game A, and check it looks like this:
<img width="1066" alt="Screenshot 2025-05-15 at 10 49 17 PM" src="https://github.com/user-attachments/assets/cb903cea-128d-4afd-bf6f-ea5ef8aaea8a" />


> AI Reference: Learn how to update published/unpublished status (modify a data object and integrate this with html form) from ChatGPT, and modify the example code for our use.

Commit:
[knowledge-base] WIP: moderator single game

- add basic view/url for both game list page and single game page
- roughly complete the game list page (displays a list of games and their
    manage page for moderator)

For: #794

Commit:
[knowledge-base] moderator backend single game

- add view for single game
- update html/url accordingly
- allow publish/unpublish/switch by adding a ModeratorSetPublishedGuide view

For: #794

Commit:
[knowledge-base] fix moderator backend single game

- forget to filter status=1 in previous commit in SingleGame page
- forget to add a display for game with no approved guides, or for no games
 in GameList page
Addressed these minor issues in this commit.

For: #794

Luca-W-D commented on src/chigame/knowledge_base/views.py:
```diff
@@ -155,7 +156,7 @@ def get_queryset(self):
 def DownloadGuide(request, pk):
     guide = get_object_or_404(Guide, pk=pk)
-    if guide.author != request.user:
+    if guide.author != request.user and not request.user.moderator:
```
Nice catch

Luca-W-D commented on src/chigame/knowledge_base/views.py:
```diff
@@ -246,3 +247,71 @@ def post(self, request, *args, **kwargs):
+        game = get_object_or_404(Game, pk=pk)
+        queryset = Guide.objects.filter(game_id=game, status=1)
+
+        # for sorting
```
This code appears to be duplicated

Luca-W-D commented on src/chigame/knowledge_base/views.py:
```diff
@@ -246,3 +247,71 @@ def post(self, request, *args, **kwargs):
+            queryset = queryset.order_by("recent_upload")
+        else:  # default: newest first
+            queryset = queryset.order_by("-recent_upload")
+        # for sorting
```
The `sort` code appears to be duplicated

yizou275 replied:
Good catch! Thanks!

Luca-W-D reviewed (approved):
Great work! I followed the tests you described and it appears to work as expected.

1. Yep, these paths exist and display as shown
2. The `view` link works for all included games. For the Go game guide, only the guide with ID=2 is shown, as you describe
3. After unpublishing, the moderator view shows N/A as the current guide ID; it also no longer appears on the landing page. Works great!
4. The "manage chess page guide" updates quite nicely when you swap which guide is published. Great, responsive UX. I also checked that the actual published version changes too.
5. The sorting works as expected, but (although this is nitpicky) the phrase "Sort by:" is current included part of a value in the list. Therefore, when you switch to "Oldest update", it no longer says "Sort by:"
<img width="376" alt="image" src="https://github.com/user-attachments/assets/b5515501-9c1d-4595-8703-69c0edf58fb8" />
<img width="389" alt="image" src="https://github.com/user-attachments/assets/2b5c7a4c-9cae-49da-8566-b8c1a1d38838" />

6. Yep– matches the image.

Overall, I think this is a great feature– definitely will be included in our final presentation! The only things to note are the duplicated (I think) code in `knowledge_base/views.py:278-289` and "sort by" values (item 5 above). Both are inconsequential/stylistic, so I'll go ahead and approve. Thanks, Yi!

Commit:
[knowledge-base] fix moderator single game page

- remove duplicate code for sorting
- add "Sort by" in the "Oldest update" sorting option, and change
  "Search" to "Sort"
- remove migration file that doesn't belong to this PR

For: #1043

yizou275 commented:
Thanks Luca! Removed duplicate code and changed sorting value.
Also changed minor wording ("search"->"sort", "TimeStamp"->"Timestamp") on the page, and removed unrelated migration file in the recent commit (the migration file for contributor manage guide page wasn't uploaded initially, but I notice in David's PR#1010 he needs to update the migration file 0004 anyway, so I remove it from this PR). 
I'll ask for Jacq's review now. Thank you for your help!

yizou275 committed:
Merge branch 'dev' into knowledge-base/moderator/single-game-management-backend

jacqsmei committed:
Merge branch 'dev' into knowledge-base/moderator/single-game-management-backend

jacqsmei reviewed (approved):
nice work on this!

jacqsmei commented:
Issue Score: Excellent

Comments:
Awesome work! Nice PR description and teamwork on the review

<!-- {"score": "excellent", "minor": false} -->
